src/directory.rs

```diff
@@ -199,6 +200,148 @@ impl Directory {
+                            stack.push((diff_entry, path.join(dir_name)))
+                        }
+                        for dir_name in diff.deleted.clone() {
+                            go_deleted(path.join(dir_name).as_path(), f).unwrap();
+                        }
+                    }
+                }
+            }
+            Ok(())
+        }
+        fn go_deleted(path: &Path, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            writeln!(f, "D {}", path.to_str().unwrap())
+        }
+
+        for (path, dir_entry) in self.added.clone() {
+            go_added(&PathBuf::from(path), &dir_entry, f).unwrap();
+        }
+        for (path, diff_entry) in self.modified.clone() {
+            go_modified(&PathBuf::from(path), &diff_entry, f).unwrap();
+        }
+        for path in self.deleted.clone() {
+            go_deleted(&PathBuf::from(path), f).unwrap();
+        }
+        Ok(())
+    }
+}
+
+#[test]
+fn test_diff_display() {
+    let diff_empty: Diff = Diff {
+        deleted: BTreeSet::new(),
+        added: BTreeMap::new(),
+        modified: BTreeMap::new(),
+    };
+    assert_eq!(diff_empty.to_string(), "");
+
+    let deleted_foo = BTreeSet::from([String::from("foo")]);
+    let added_bar: BTreeMap<String, DirectoryEntry> = vec![(
+        String::from("bar"),
+        DirectoryEntry::File(ObjectId::from(&vec![])),
+    )]
+    .into_iter()
+    .collect();
+
+    let diff_1: Diff = Diff {
+        deleted: BTreeSet::new(),
+        added: added_bar.clone(),
+        modified: BTreeMap::new(),
+    };
+    assert_eq!(diff_1.to_string(), "A bar\n");
+
+    let diff_2: Diff = Diff {
+        deleted: deleted_foo.clone(),
+        added: BTreeMap::new(),
+        modified: BTreeMap::new(),
+    };
+    assert_eq!(diff_2.to_string(), "D foo\n");
+
+    let diff_3: Diff = Diff {
+        deleted: deleted_foo.clone(),
+        added: added_bar.clone(),
+        modified: BTreeMap::new(),
+    };
+    assert_eq!(diff_3.to_string(), ["A bar", "D foo", ""].join("\n"));
+
+    let diff_4: Diff = Diff {
+        deleted: deleted_foo.clone(),
+        added: added_bar.clone(),
+        modified: vec![(
+            String::from("baz"),
+            DiffEntry::File(ObjectId::from(&vec![])),
+        )]
+        .into_iter()
+        .collect(),
+    };
+    assert_eq!(
+        diff_4.to_string(),
+        ["A bar", "M baz", "D foo", ""].join("\n")
+    );
+
+    let diff_5: Diff = Diff {
+        deleted: deleted_foo.clone(),
+        added: added_bar.clone(),
+        modified: vec![
+            (String::from("a"), DiffEntry::Directory(Box::new(diff_2))),
+            (String::from("baz"), DiffEntry::Directory(Box::new(diff_4))),
+        ]
+        .into_iter()
+        .collect(),
+    };
+    assert_eq!(
+        diff_5.to_string(),
+        [
+            "A bar",
+            "D a/foo",
+            "A baz/bar",
+            "D baz/foo",
+            "M baz/baz",
+            "D foo",
+            ""
```

Comment:
This is arbitrarily ordered. Maybe it could be nice to have an intermediate representation of the `Diff` as list of `Path`s paired/tagged with a `Deleted`/`Modified`/`Added` enum, and could sort by `Path` before displaying? Attempted the simplest thing for now.

Reply:
Yeah, that might be nice. I think that grouping makes sense to view here, but don't feel any pressure, we can merge it as is.